### PR TITLE
Minimalist client side port

### PR DESCRIPTION
**Pull request by dsathyakumar**
## Description
A client side port of `@ebay/retriever` without the dependency of `lodash.get` & `lodash.memoize`

## Context
- For minimalist use on the client side
- retains the original functions of `@ebay/retriever` and 
- re-maps via the `browser` prop.
- claws through nested data safely.

@yomed @sidsakhadeo

### CODE REVIEW

**Comment by sidsakhadeo** [comment]
@dsathyakumar Can you publish some numbers about the performance on client without `lodash`?

File: browser.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    return type;
+}
+
+/**
+ * Normalizes the path String sent, to array types
+ * @param {String} path
+ * @returns {String}
+ */
+function normalize(path) {
+    if (typeof path === 'string') {
```
Comment (sidsakhadeo) [inline browser.js]: I'm not sure why this `if` condition would be necessary

File: browser.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
+'use strict';
+/* **
+ * @description A client side port of `@ebay/retriever`
```
Comment (sidsakhadeo) [inline browser.js]: (minor) I think you can remove this comment block. This file is self-explanatory.

File: browser.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    result = !(typeofResult === 'undefined' || typeofResult === 'null');
+    return result;
+}
+
+/**
+ * Exported Function to claw through nested data safely
+ * Just returns the value without logging
+ * @param {Object} obj
+ * @param {String} path
+ * @param {*} defaultVal - if nothing provided, defaulst to ''
```
Comment (sidsakhadeo) [inline browser.js]: (minor) `defaults` spelling

File: browser.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+function get(obj, path, defaultVal) {
+    return access(obj, path, defaultVal);
+}
+
+/**
+ * Exported Function to claw through nested data safely
+ * Logs the value & returns it
+ * @param {Object} obj
+ * @param {String} path
+ * @param {*} defaultVal - if nothing provided, defaulst to ''
```
Comment (sidsakhadeo) [inline browser.js]: (minor) `defaults` spelling

**Review by sidsakhadeo — CHANGES_REQUESTED** [review]

**Comment by sidsakhadeo** [comment]
@dsathyakumar @yomed I think we should have the same `test` and `perf.js` run for both server and client side code without actually duplicating the code for these in separate files. What do you think?

**Comment by yomed** [comment]
@dsathyakumar Thanks for the PR, but I think we need to consider if we actually want to take this type of change. There is a lot of value in using lodash, such as 1) robustness, 2) reduced maintenance, 3) identical functionality across client and server, etc. As @sidsakhadeo mentioned, can you give more details on the benefits of this approach in general? Then if we do want to move forward, we can dig into the implementation details.

**Comment by sidsakhadeo** [comment]
@dsathyakumar Also, how much is the gain in dropping `lodash` on the client side?

**Comment by dsathyakumar** [comment]
Thanks @yomed @sidsakhadeo 
Before the change it was 6.2kb gzipped and now its 1.9kb gzipped.
The re-map was done to cut bytes, however, the same change could be used on the server too (thereby maintaining code uniformity). The re-map could be removed. IIRC, initial queries are fast with this change, but lodash's memoization makes repeated querying quicker.

**Comment by sidsakhadeo** [comment]
> but lodash's memoization makes repeated querying quicker

I think this is useful for us. In most cases, we are querying subsections of a huge object to render templates. If repeated querying is slower, this needs to be addressed.

**Comment by sidsakhadeo** [comment]
@dsathyakumar I'm guessing in the previous comment with the benchmarks, the `old` means the server side retriever calls with `lodash` and the `new` means the client side retriever calls without `lodash`. Is that correct?

**Comment by yomed** [comment]
A few thoughts:
1) I'd like to avoid issues that may arise from having different implementations across server/client, especially for a case like this where it's simple to have an isomorphic solution.
2) Given the above, we should pick a single isomorphic implementation, and weigh the tradeoffs of various solutions in doing so.
3) Using an existing solution has several advantages in terms of durability and maintenance, while also allowing us to focus on the parts that make `retriever` unique (type checking, optional logging, and convenient defaults for template rendering).
4) Given that, I'm open to using other existing libraries if we believe there may be significant advantages in doing so. But I'll also say that using Lodash for this has served us well so far.

**Comment by dsathyakumar** [comment]
ok! @yomed & @sidsakhadeo closing this then.
Thanks for taking time to review this 👍

**Pull request closed without merging**